artiq/applets/big_number.py

```diff
@@ -26,54 +30,70 @@ def __init__(self, args, ctl):
         QtWidgets.QStackedWidget.__init__(self)
         self.dataset_name = args.dataset
         self.ctl = ctl
+        self.value = None
+        self.metadata = {}
 
-        self.lcd_widget = QResponsiveLCDNumber()
-        self.lcd_widget.setDigitCount(args.digit_count)
-        self.lcd_widget.doubleClicked.connect(self.start_edit)
-        self.addWidget(self.lcd_widget)
+        font = QtGui.QFont()
+        font.setPointSize(60)
+
+        self.display_widget = QResponsiveLabel()
+        self.display_widget.setFont(font)
+        self.display_widget.setAlignment(
+            QtCore.Qt.AlignRight | QtCore.Qt.AlignVCenter)
+        self.display_widget.doubleClicked.connect(self.start_edit)
+        self.addWidget(self.display_widget)
 
         self.edit_widget = QCancellableLineEdit()
-        self.edit_widget.setValidator(QtGui.QDoubleValidator())
-        self.edit_widget.setAlignment(QtCore.Qt.AlignRight)
-        self.edit_widget.editCancelled.connect(self.cancel_edit)
+        self.edit_widget.setFont(font)
+        self.edit_widget.setAlignment(
+            QtCore.Qt.AlignRight | QtCore.Qt.AlignVCenter)
+        self.edit_widget.editCancelled.connect(
+            lambda: self.setCurrentWidget(self.display_widget))
         self.edit_widget.returnPressed.connect(self.confirm_edit)
         self.addWidget(self.edit_widget)
 
-        font = QtGui.QFont()
-        font.setPointSize(60)
-        self.edit_widget.setFont(font)
+        float_regexp = r"(([+-]?\d+(\.\d*)?|\.\d+)([eE][+-]?\d+)?)"
+        self.regexp = "^" + float_regexp + r"( [a-zA-Z]*)?$"
+        self.edit_widget.setValidator(
+            QtGui.QRegExpValidator(QtCore.QRegExp(self.regexp)))
 
-        self.setCurrentWidget(self.lcd_widget)
+        self.setCurrentWidget(self.display_widget)
 
     def start_edit(self):
-        # QLCDNumber value property contains the value of zero
-        # if the displayed value is not a number.
-        self.edit_widget.setText(str(self.lcd_widget.value()))
+        if self.value is None:
+            self.edit_widget.setText("---")
+        else:
+            unit = self.metadata.get("unit", "")
+            scale = getattr(units, unit, 1)
```

Comment:
Ignorant of any scale param provided in metadata... will not work on any custom defined units. Let's say I defined an arbitrary unit 't' (name irrelevant) and set the scale to 1000 in browser, this will display 1000x the value in the applet TextEdit as a consequence, and I will lose the scale metadata when opening in dashboard. Use `scale_from_metadata` to handle this instead.

Reply:
I know we have gone back and forth on this, but the problematic thing with using `scale_from_metadata` mentioned in previous comments was also not passing the full metadata obj (only unit). Let's try using the `scale` metadata in this widget, should it be defined, and only override in case of a change of units (in which case only defaults should be permissable IMO). Alternatively we may lock-in the units and make them display only in this widget.